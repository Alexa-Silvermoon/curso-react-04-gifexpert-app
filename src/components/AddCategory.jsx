import { useState } from "react"
import  PropTypes  from 'prop-types';

// export const AddCategory = ({ setCategories }) => {
export const AddCategory = ({ onNewCategory }) => {

    const [inputValue, setinputValue] = useState('');

    // const onInputChange = (event) => {
    const onInputChange = ({ target }) => { //para que mi input sea dinamico

        // console.log( event.target.value );
        // console.log( target.value ); //muestra letra por letras a medida que voy escribiendo

        // setinputValue( event.target.value );
        setinputValue( target.value );
    }

    const onSubmit = ( event ) => { //para la tecla enter

        event.preventDefault(); // evita que el navegador se refresque

        if ( inputValue.trim().length <= 1 ) return; //debe haber mas de un caracter para realizar la busqueda

        // console.log( event ); // muestra: yntheticBaseEvent {_reactName: 'onSubmit', _targetInst: null, type: 'submit', nativeEvent: SubmitEvent, target: form, …}
        // console.log('enter: ',inputValue ); //este valor lo enviare para hacer busquedas despues
        
        // setCategories( ( categories ) => [ inputValue, ...categories ]);
        setinputValue(''); // limpia la caja de texto del input despues del enter
        onNewCategory( inputValue.trim() ); //trim() elimina espacios en blanco innecesarios en el input
    }

  return (

    // <form onSubmit={ ( event ) => onSubmit( event )}>
    // aria-label se usa para AddCategory.test.jsx
    <form onSubmit={ onSubmit } aria-label="form">

        <input
            type="text"
            placeholder="Buscar gifs"
            value={ inputValue }
            // onChange={ (event) => onInputChange(event) }
            onChange={ onInputChange }
        ></input>

    </form>

  )
}

AddCategory.propTypes = {

    onNewCategory: PropTypes.func.isRequired //func de function
}