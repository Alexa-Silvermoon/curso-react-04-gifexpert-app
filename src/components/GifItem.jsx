
// export const GifItem = ( props ) => {
// export const GifItem = ( { image } ) => {
export const GifItem = ( { title, url, id } ) => {

  // console.log( props );
  // console.log( image );
  // console.log( { title, url, id } );

  // finalmente renderiza los gif en el html
  return (

    <div className="card">

      <img src={ url } alt={ title }></img>
      <p>{ title }</p>

    </div>
  )
}
 // https://www.udemy.com/course/react-cero-experto/learn/lecture/19758956#questions