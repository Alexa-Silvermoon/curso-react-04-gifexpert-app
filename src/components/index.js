
export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';
// archivo de barril, tambien conocido como archivo index