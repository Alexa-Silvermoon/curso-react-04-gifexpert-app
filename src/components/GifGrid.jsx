// import { useEffect, useState } from "react";
import { GifItem } from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";
// import { getGifs } from "../helpers/getGifs";

export const GifGrid = ( { category } ) => {
  
  // getGifs(category);

  const { images, isLoading } = useFetchGifs( category ); //custom hook

  // console.log( { images, isLoading } );

  // el siguiente bloque de codigo lo he metido en: useFetchGifs
  /* const [images, setImages] = useState( [] );

  const getImages = async() => {

    const newImages = await getGifs( category );
    setImages( newImages );

  }

  useEffect(() => { // useEffect no se puede usar con async y await, por esa razon se creo la funciones getImages aparte

    getImages();
    
  }, []) */
  
  return (

    <>
        <h3> {category} </h3>
        {
          /* isLoading? // forma 1
          ( <h2>Cargando...</h2> )
          : null // el null no se renderiza en react */

          isLoading && ( <h2>Cargando...</h2> ) // forma 2
        }

            <div className="card-grid">
              {
    
                /* images.map( ( { id, title} ) => ( //opcion 3
    
                  <GifItem key={ id }></GifItem>
    
                )) */

                /* images.map( ( image  ) => ( //opcion 4
    
                  <GifItem 
                          key={ image.id }
                          image={ image.title }
                          url={ image.url }></GifItem>
    
                )) */

                images.map( ( image  ) => ( //opcion 5 con el spread ... hago exparsir la image, es decir sus propiedades
    
                  <GifItem 
                          key={ image.id }
                          { ...image }></GifItem>
    
                ))
    
              }
    
            </div>

        {/* <ol> */}

          {

            /* images.map( ( img ) => ( //opcion 1

              <li key={ img.id } > { img.title } </li>

            )) */

            /* images.map( ( { id, title} ) => ( //opcion 2

              <li key={ id } > { title } </li>

            )) */

          }

        {/* </ol> */}

    
    </>
  )
}
