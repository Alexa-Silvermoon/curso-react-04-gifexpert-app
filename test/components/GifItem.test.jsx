import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

// https://www.udemy.com/course/react-cero-experto/learn/lecture/19784000#questions

/* 
GifItem.propTypes = {

  title: propTypes.string.isRequired,
  url: propTypes.string
  
}
 */

describe('Pruebas en GifItem.jsx', () => {

    const title = 'string title';

    // const url = 'https://media3.giphy.com/media/3o6fJ5z2bgCLBshZUA/giphy.gif?cid=e9ea52aa1b2dzwzaxtybn4altkqgz5xju7e0aipikr4bssog&rid=giphy.gif&ct=g';
    const url = 'https://ejemplo.com/imagen.jpg';

    test('Evaluar el snapshot, title obligatorio con propTypes', () => {

        const { container} = render(<GifItem title={title} url={url}></GifItem>);

        expect( container ).toMatchSnapshot();

    });

    test('debe mostrar la imagen con el URL y el ALT indicado', () => {

        const { container} = render(<GifItem title={title} url={url}></GifItem>);
        screen.debug();

        // probando si hay una imagen en el url
        // expect( screen.getByRole('img').src ).toBe( url );
        
        // probando si hay una imagen en el url ( optimizado ):
        const { src, alt } = screen.getByRole('img')
        expect( src ).toBe( url ); // const url = 'https://ejemplo.com/imagen.jpg';
        expect( alt ).toBe( alt ); // const title = 'string title';

        /* 
        <body>
            <div>
                <div
                class="card"
                >
                <img
                    alt="string title"
                    src="https://ejemplo.com/imagen.jpg"
                />
                <p>
                    string title
                </p>
                </div>
            </div>
        </body>
        */

    });

    test('debe mostrar el titulo en el componente', () => {

        render( <GifItem title={title} url={url}></GifItem>);
        expect( screen.getByText( title) ).toBeTruthy();

    });

});