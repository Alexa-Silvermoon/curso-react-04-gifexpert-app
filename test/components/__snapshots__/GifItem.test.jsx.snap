// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifItem.jsx Evaluar el snapshot, title obligatorio con propTypes 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="string title"
      src="https://ejemplo.com/imagen.jpg"
    />
    <p>
      string title
    </p>
  </div>
</div>
`;
